vllm_spyre/v1/worker/spyre_model_runner.py

```diff
@@ -305,38 +305,103 @@ def prepare_model_input(
             (input_tokens, input_positions,
              input_masks) = self._prepare_decode(
                  scheduler_output.scheduled_cached_reqs)
-            # seq_lens = []
-            num_reqs = len(scheduler_output.scheduled_cached_reqs)
-
-        # TODO: Build the rest of the SamplingMetadata correctly
-        dummy_tensors = lambda v: torch.full(
-            (num_reqs, ), v, device=self.device)
-        dummy_metadata = SamplingMetadata(
-            temperature=dummy_tensors(0.0),
-            all_greedy=False,
-            all_random=False,
-            top_p=None,
-            top_k=None,
-            min_p=None,
-            generators={},
-            max_num_logprobs=self._max_logprobs,
-            no_penalties=True,
-            prompt_token_ids=None,
-            frequency_penalties=dummy_tensors(0.1),
-            presence_penalties=dummy_tensors(0.1),
-            repetition_penalties=dummy_tensors(0.1),
-            output_token_ids=[[] for _ in range(num_reqs)],
-            min_tokens={},
-            logit_bias=[None for _ in range(num_reqs)],
-            allowed_token_ids_mask=None,
-        )
+
+        sampling_metadata = self.input_batch.sampling_metadata
 
         return ModelInputForSpyre(input_tokens=input_tokens,
                                   input_positions=input_positions,
                                   input_masks=input_masks,
-                                  sampling_metadata=dummy_metadata,
+                                  sampling_metadata=sampling_metadata,
                                   is_prompt=is_prompt)
 
+    def _update_states(self, scheduler_output: "SchedulerOutput") -> None:
+        """Update the cached states and the persistent batch with the scheduler
+        output.
+
+        The SamplingMetadata is updated and copied to the device if there is a
```

Comment:
nit: I don't think we're actually doing any "copying to device" operations here, right?